static/_posts/2024-04-01-vgaterm-intro-001.md

```diff
@@ -0,0 +1,171 @@
+
+To solve the speed issue we utilized the esp32c3 quad [SPI](https://en.wikipedia.org/wiki/Serial_Peripheral_Interface) module allowing 4 bits to be clocked out at up to 80 MHz - much faster than bit-banging. Our actual implementation ended up using a 40 MHz SPI clock speed.
+
+Lastly, even with the faster SPI pixel output, emitting the full 640x480 frame proved to take too much time of the frame, so we reduced the resolution to 640x400. For a whole frame at 40MHz per half-byte this would take 15.36 ms, giving us only 1.3 ms per 60 Hz frame to do non-display work. At the smaller size it only takes 12.8 ms to emit a frame and 3.8 ms leftover time. Plus, 640x400 is widescreen, so it looks more modern!
+
+With these pieces we could design a circuit that collects two 4 bit [nibbles](https://en.wikipedia.org/wiki/Nibble) into a byte (representing the color of a single pixel) and load it into a FIFO memory. At the same time, the VGA timing hardware would read pixel bytes exactly when needed.
+
+### Logical Components
+
+There are five main components to Vgaterm:
+
+* VGA Timing
+* CPU Pixel Output
+* Displaying a Pixel Byte
+* Digital to Analog Converters (DAC)
+* IO control and video buffer updates
+
+{% include figure.html image="/assets/images/vgaterm-logical-diagram.jpg" caption="Vgaterm logical diagram. Details were omitted to show the high level units" width=800 %}
+
+Repositories containing all our code and schematics:
+* esp32c3 code: [https://github.com/rustbox/esp32c3-vgaterm](https://github.com/rustbox/esp32c3-vgaterm)
+* VGA timing generation code and Programmable Logic Device code: [https://github.com/rustbox/cuddly-robot](https://github.com/rustbox/cuddly-robot)
+* PCB schematics and CAD: [https://github.com/rustbox/vgaterm-schematics](https://github.com/rustbox/vgaterm-schematics)
+
+### VGA Timing
+
+VGA comes in several modes, and we target the 640x480, 60Hz frame rate mode - the base standard. VGA has 3 analog signals between 0 and 0.7 volts representing the amount of Red Green and Blue per pixel. There are also two timing signals `H_SYNC` and `V_SYNC` that tell the monitor what resolution mode to be in and when to sample the color signals to actually display a pixel. When the monitor is actively sampling colors and displaying pixels we say that this portion of the frame is `visible`. There is time during a complete frame when this isn't the case and this is the _blanking area_. Lastly, the _pixel clock_ speed is the rate at which pixels are shown in the visible portion, and in our case that's 25.175 MHz. In the diagram above, I've labeled a signal `CLK_VIS`, and this is the _pixel clock_ signal during the `visible` time. To see detailed VGA timing information, see [tiny vga](http://tinyvga.com/vga-timing/640x480@60Hz).
```

Comment:
If you really want to introduce blanking area and define H_SYNC and V_SYNC (which, I don't think you need to), then I'd suggest taking it from the other end: why did we do the timing signals first?  I remember being really glad we did that, but the reader doesn't know why. What happened the handful of times something went wrong there?

Reply:
FWIW, I still think this paragraph feels out-of-place: we were just doing high-level block diagrams, and now you're telling me about pixel clocks to 3 decimal places of precision? what gives? worse, it's the 2nd (actually 3rd) time the reader is encountering this detail, and there's not really a payoff for it: we never talk about other modes, or what would need to change to support them, so "pixel clock" is just as specific as "25.175MHz pixel clock" (for the purposes of this post, anyway).

Your next paragraph covers everything that needs to be said really well, I think, so I'd still suggest striking it out; failing that, maybe move it to the bottom of this section so it's not quite such an abrupt level-of-detail transition?